### PR TITLE
Unicode identifiers (PEP 3131)

### PR DESCRIPTION
**Pull request by da-woods**
Issue: https://github.com/cython/cython/issues/2601

I'm sure this isn't 100% complete and bug-free but it basically seems to work and doesn't look to break the tests (on my PC in Python 3 at least).

There's a few things you may want to query:
1. I wonder if the changes to the lexer lexicon.py should be switchable by environment or confirguation options - at the moment they're permanently on.
2. I'm also not completely convinced all the changes are done in the right place - I had to scatter them around a bit between different levels (mostly symtab and nodes, but also one in vistor) which seems slighly illogical.
3. I think it is possible to get name collisions (if you really try) - the name mangling doesn't distinguish between unicode in a class name or in a function name or both. It may not be a huge concern but possibly deserves a closer look

### CODE REVIEW

File: Cython/Compiler/Lexicon.py [hunk 1]
```diff
@@ -143,3 +546,65 @@ def prefixed_digits(prefix, digits):
...
+    plex_ranges = _make_split_strings(plex_ranges)
+    return "Any(({0}).decode('utf-8')) | Range(({1}).decode('utf-8'))".format(
+        single_chars,plex_ranges)
+
+def _make_split_strings(chars,splitby=20):
+    chars = "".join(chars).encode('utf-8')
+    out = []
+    for i in range(0,len(chars),splitby):
+        out.append(repr(chars[i:i+splitby]))
+    return "+\n".join(out)
```
Comment (scoder) [inline Cython/Compiler/Lexicon.py]: No need for a `+` here, just let the parser join the strings.

File: Cython/Compiler/Nodes.py [hunk 2]
```diff
@@ -1895,7 +1906,8 @@ def generate_function_definitions(self, env, code):
         if use_refnanny:
             tempvardecl_code.put_declare_refcount_context()
             code.put_setup_refcount_context(
-                self.entry.name, acquire_gil=acquire_gil_for_refnanny_only)
+                self.entry.name,
+                acquire_gil=acquire_gil_for_refnanny_only)
```
Comment (scoder) [inline Cython/Compiler/Nodes.py]: Worth reverting this change.

File: Cython/Compiler/Nodes.py [hunk 3]
```diff
@@ -879,6 +883,13 @@ class CArgDeclNode(Node):
     annotation = None
     is_dynamic = 0
 
+    @property
+    def name_cstring(self):
```
Comment (scoder) [inline Cython/Compiler/Nodes.py]: Can probably be avoided by avoiding the `AttributeError`.

File: Cython/Compiler/Visitor.py [hunk 4]
```diff
@@ -659,6 +660,7 @@ def _dispatch_to_handler(self, node, function, arg_list, kwargs):
     def _dispatch_to_method_handler(self, attr_name, self_arg,
                                     is_unbound_method, type_name,
                                     node, function, arg_list, kwargs):
+        type_name = punycodify_name(type_name) # FIXME: this seems like it should be handled much earlier
```
Comment (scoder) [inline Cython/Compiler/Visitor.py]: Why is this needed?

File: Cython/Compiler/Symtab.py [hunk 5]
```diff
@@ -42,6 +42,22 @@ def c_safe_identifier(cname):
         cname = Naming.pyrex_prefix + cname
     return cname
 
+def punycodify_name(cname):
+    # modified from  PEP489
+    try:
+        cname = cname.encode('ascii')
+    except UnicodeEncodeError:
+        cname = cname.encode('punycode').replace(b'-', b'_')
+        if cname.startswith(Naming.pyrex_prefix.encode('ascii')):
```
Comment (scoder) [inline Cython/Compiler/Symtab.py]: I would rather like to see punicode identifiers use a different prefix. For example, we use `__pyx_v_` for variables. Why not use `__pyx_vp_` for punicode variable names?

File: Cython/Compiler/Nodes.py [hunk 6]
```diff
@@ -3393,7 +3405,7 @@ def generate_function_definitions(self, env, code):
                 self.return_type.declaration_code(Naming.retval_cname),
                 retval_init))
         code.put_declare_refcount_context()
-        code.put_setup_refcount_context('%s (wrapper)' % self.name)
+        code.put_setup_refcount_context(EncodedString('%s (wrapper)' % self.name))
```
Comment (scoder) [inline Cython/Compiler/Nodes.py]: Could be that something like this is needed in a couple of more places – might be one of the reasons why the `AttributeError`s appeared for you.

File: Cython/Compiler/Lexicon.py [hunk 7]
```diff
@@ -26,6 +26,408 @@ def make_lexicon():
     hexdigit = Any("0123456789ABCDEFabcdef")
     indentation = Bol + Rep(Any(" \t"))
 
+    # Some functions at the end of the file show how to generate these strings.
+    # They're pretty slow to generate at runtime, and require Python3, so
+    # are just included directly here
+    unicode_start_character = (
```
[the remark this answers is not in the document]
Comment (da-woods) [reply Cython/Compiler/Lexicon.py]: I can't find an interface that lets you request all the characters in category. It's certainly possible to iterate character-by-character and use `unicodedata` to look up the category. However both this and my current approach of "iterate through strings trying `isidentifier`) are pretty slow - about 0.5s on my PC. Therefore I don't think this code should be run every time Cython starts, hence the huge constants.

I agree that if a better way exists it'd be good to use it!

File: Cython/Compiler/Symtab.py (hunk 5 above)
Comment (da-woods) [reply Cython/Compiler/Symtab.py]: The line after this snippet does that, so a variable becomes `__pyx_Uv_`; a function becomes `__pyx_Uf`. It does that just by inserting the `U` rather than adding more prefixes `Naming.py`.

(Where that might cause problems is on unicode attributes in unicode classes - only one `U` gets inserted so it isn't clear which part of the identifier is unicode. I think if you're really unlucky you could end up with a name collision, but it's pretty unlikely)

**Comment by da-woods** [comment]
I'll go through the rest of your comments (the ones which require changing/looking at the code) in the next couple of days....

File: Cython/Compiler/Symtab.py (hunk 5 above)
Comment (scoder) [reply Cython/Compiler/Symtab.py]: Then, let's make that an official `pyunicode_identifier_prefix` in `Naming.py`, so that it's visible where the prefix comes from.

File: Cython/Compiler/Lexicon.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (scoder) [reply Cython/Compiler/Lexicon.py]: Thinking some more, since `unicodedata` is regularly updated with new CPython releases, and might learn new identifier characters over time, we should keep this set of characters independent of the specific CPython version that Cython runs with. Meaning, this set should be generated in a controlled way, always with the latest CPython version, and then committed.

BTW, since this is Python code, and the string is generated and guaranteed to consist only of printable characters, why not leave the character decoding to the Python parser and use the real Unicode characters instead of byte strings? (That would make `Lexicon.py` a UTF-8 file, but I'm fine with that.)

[the remark this answers is not in the document]
Comment (scoder) [reply Cython/Compiler/Lexicon.py]: Also, I would prefer having the identifier character strings as global constants in the module, after the `make_lexicon()` function. That also simplifies their generation a bit. You can use start+end markers like `# GENERATED CODE STARTS HERE` to make them easy to update programmatically. Preferably by a separate script in `bin/`. And if you want to be really ambitious, then you can add a version marker of the CPython version that was used to generate the strings, and let the script print a big warning when they are regenerated with an older version. :)

**Comment by da-woods** [comment]
I've uploaded a commit that implements your suggestions for generating Lexicon.py. I've marked those comments as "resolved" but obviously feel free to query it further if needed.

File: Cython/Compiler/Visitor.py (hunk 4 above)
Comment (da-woods) [reply Cython/Compiler/Visitor.py]: I think I'd misidentified the problem slightly. _get_handler does `getattr` with a unicode string. This isn't valid when running Cython in Python 2 since only ascii strings are allowed. I've adopted a different approach of returning None from _get_handler if it's passed a unicode string. I'm not sure I understand the purpose of handlers enough to know if this is always valid, but it shouldn't break any existing code (since existing code should be passing unicode strings)

File: Cython/Compiler/StringEncoding.py [hunk 8]
```diff
@@ -138,6 +138,13 @@ def contains_surrogates(self):
     def as_utf8_string(self):
         return bytes_literal(self.utf8encode(), 'utf8')
 
+    def as_encoded_c_string_literal(self):
```
[the remark this answers is not in the document]
Comment (da-woods) [reply Cython/Compiler/StringEncoding.py]: I've changed the name of this to be "as_c_string_literal" which matches what's already in BytesLiteral - on reflection that makes more sense than adding a new interface.

**Comment by da-woods** [comment]
I've added a commit which hopefully addresses of the rest of your comments. I think it passes all the tests (at least running in Python 3), but I wouldn't be surprised if it breaks some obscure, currently untested existing code. I was considering (but haven't implemented) replacing  `name.as_c_string_literal()` with a global function `as_c_string_literal(name)` which could at least catch the exception and print a message to identify strings that have been missed instead to just failing.

I've also added a check for language_level since unicode identifiers aren't supported in Python2. (Ideally this should also be accompanied by a C-compile-time too, but I don't think this is a priority since it'd entirely be the user's fault, and all they'll end up with is some variables that they can't access)

File: bin/cython-generate-lexicon.py [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+        raise ValueError('argv: "{0}" not recognised', sys.argv[1])
+
+    print("Writing to file", outfile)
+    with open(outfile,'w') as f:
+        f.write(output.getvalue())
+
+
+# The easiest way to generate an appropriate character set is just to use the str.isidentifier method
+# An alternative approach for getting character sets is at https://stackoverflow.com/a/49332214/4657412
+def get_start_characters_as_number():
```
Comment (scoder) [inline bin/cython-generate-lexicon.py]: Feel free to use `@functools.lru_cache()` to avoid creating this list more than once.

File: bin/cython-generate-lexicon.py [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+    #  a list pairs of characters representing closed ranges
+    char_num_list = sorted(char_num_list)
+    first_good_val = char_num_list[0]
+
+    single_chars = []
+    ranges = []
+    for n in range(1,len(char_num_list)):
+        if char_num_list[n]-1 != char_num_list[n-1]:
+            # discontinuous
+            if first_good_val==char_num_list[n-1]:
```
Comment (scoder) [inline bin/cython-generate-lexicon.py]: Please always keep spaces around these operators, especially in non-trivial expressions.
https://www.python.org/dev/peps/pep-0008/#other-recommendations

File: Cython/Compiler/Symtab.py [hunk 11]
```diff
@@ -446,6 +463,19 @@ def declare(self, name, cname, type, pos, visibility, shadow = 0, is_type = 0, c
...
+            ll2 = self.context.language_level == 2
+        except AttributeError:
+            # some scopes, for example the builtin scope, don't
+            # seem to have a context - don't worry about these ones
+            ll2 = False
+        if ll2 and name:
+            try:
+                name.encode('ascii')
+            except UnicodeEncodeError:
+                error(pos, "Unicode identifiers are only supported at language_level>=3")
```
[the remark this answers is not in the document]
Comment (da-woods) [reply Cython/Compiler/Symtab.py]: No - not really. I've removed the check.

**Comment by da-woods** [comment]
OK - I think all your comments are addressed. I've added some run-time tests as suggested (slightly hackily since they only work on Python 3, but the build tests shouldn't care which version of Python you're using - I couldn't see a better way of doing this).

File: bin/cython-generate-lexicon.py [hunk 12]
```diff
@@ -26,10 +27,12 @@
 from Cython.Compiler import Lexicon
 
 def main():
-    if (len(sys.argv) != 2 or
-        (sys.argv[1] not in ['--overwrite','--here'])):
+    arg = '--overwrite'
+    if len(sys.argv) == 2:
+        arg = sys.argv[1]
+    if arg not in ['--overwrite','--here']:
```
Comment (scoder) [inline bin/cython-generate-lexicon.py]: `if len(sys.argv) > 2 or …`

File: bin/cython-generate-lexicon.py [hunk 13]
```diff
@@ -63,12 +66,12 @@ def main():
...
 
-    if sys.argv[1] == "--here":
+    if arg == "--here":
         outfile = "Lexicon.py"
-    elif sys.argv[1] == "--overwrite":
+    elif arg == "--overwrite":
         outfile = Lexicon.__file__
     else:
-        raise ValueError('argv: "{0}" not recognised', sys.argv[1])
+        raise ValueError('argument "{0}" not recognised', arg)
```
Comment (scoder) [inline bin/cython-generate-lexicon.py]: I think this should say `'…'.format(arg)`, but since this case is already handled above, a bare `else:` for the second case, maybe with an `assert`, seems enough.

File: tests/run/unicode_identifiers.pyx [hunk 14]
```diff
@@ -1,9 +1,48 @@
 # -*- coding: utf-8 -*-
 # cython: language_level=3
-# mode: compile
+# mode: run
 # tag: pep3131
 
-"ναμε1"
+# Code with unicode_identifiers can be compiled with Cython running either Python 2 or 3.
+# However it can only be run in Python 2. Therefore no doctests are run in Python2.
```
Comment (scoder) [inline tests/run/unicode_identifiers.pyx]: -> `can only be run in Python 3`

That's not entirely true, though. A compiled ASCII-named Python function that internally uses and calls non-ASCII names and functions should be usable from a Python doctest. Would be nice to have this case included (in both Py2 and Py3). If that works, it allows Cython modules also in Python 2 that can use Unicode identifiers internally, as long as they provide an ASCII-only Python-level API.

File: Cython/Compiler/ParseTreeTransforms.py [hunk 15]
```diff
@@ -1325,9 +1325,9 @@ class DecoratorTransform(ScopeTrackingTransform, SkipDeclarations):
     _properties = None
 
     _map_property_attribute = {
-        'getter': '__get__',
-        'setter': '__set__',
-        'deleter': '__del__',
+        'getter': EncodedString('__get__'),
+        'setter': EncodedString('__set__'),
+        'deleter': EncodedString('__del__'),
```
Comment (scoder) [inline Cython/Compiler/ParseTreeTransforms.py]: Ah, nice. Thanks for catching these.

File: tests/run/unicode_identifiers.pyx (hunk 14 above)
Comment (da-woods) [reply tests/run/unicode_identifiers.pyx]: I've added a few runtime tests for using Python 2 as you described. I'd be reluctant to really advertise this as a feature since it's hard to avoid any unicode leaking back to Python when it shouldn't do and there's no attempt to diagnose when the "ASCII-only Python-level API" is crossed, but practically a lot of it works.

The tests are fairly basic since it's surprisingly hard to write doctests that output unicode and work in all versions (I gave up on inspecting the traceback and locals in Python 2 for example).

**Comment by da-woods** [comment]
Updated again. While I was doing it I found that `cdef public` attributes of `cdef class` didn't work (so that's fixed). I also found that punycode sometimes generates names starting with numbers, which causes problems when they're directly used in C (for example as a struct attribute or in the vtable). I've now prefixed those (but only in the unicode case) and added some tests specifically for those names.

File: Cython/Compiler/Naming.py [hunk 16]
```diff
@@ -13,7 +13,7 @@
 
 temp_prefix       = u"__cyt_"
 
-pyunicode_identifier_prefix = (pyrex_prefix + 'U').encode('ascii')
+pyunicode_identifier_prefix = pyrex_prefix + 'U'
```
Comment (scoder) [inline Cython/Compiler/Naming.py]: I also find it much better when the constants in `Naming` are text and not (surprisingly) some of them bytes. `punicodify_name()` is the only place where bytes are handled (temporarily), everything else is text. So this is a good change.

File: Cython/Compiler/Symtab.py [hunk 17]
```diff
@@ -46,23 +46,22 @@ def punycodify_name(cname, mangle_with=None):
     # if passed the mangle_with should be a byte string
     # modified from  PEP489
     try:
-        cname = cname.encode('ascii')
+        cname.encode('ascii')
     except UnicodeEncodeError:
-        pyrex_prefix = Naming.pyrex_prefix.encode('ascii')
         cname = cname.encode('punycode').replace(b'-', b'_')
+        if not isinstance(cname,str):
+            cname = cname.decode('ascii') # for Python
```
Comment (scoder) [inline Cython/Compiler/Symtab.py]: You probably meant the comment to read "for Python 3", but I would suggest always doing this, i.e. just decode the byte string right after punycode-encoding it, also in Py2 (because, why not).

Comment (da-woods) [reply Cython/Compiler/Symtab.py]: Yes - that makes sense. I've made the change.

**Comment by scoder** [comment]
Excellent! Thanks a lot, this is a great feature to have in Cython!

**Comment by scoder** [comment]
I cleaned up the generator script a little, and then looked over the scanner again. I think normalisation isn't implemented yet, right?
https://www.python.org/dev/peps/pep-3131/#implementation

Without normalisation, combining characters won't be accepted in identifiers:
https://stackoverflow.com/questions/16467479/normalizing-unicode

That's probably ok for now (as I said, this is already a great feature as it is!), but it means that we will need to revisit the scanning of identifiers at some point.

**Comment by da-woods** [comment]
I've created a separate pull request https://github.com/cython/cython/pull/3096 that I think handles the normalization. It's fairly short and straight-forward.

I think the problem is not that combining characters aren't accepted. It's that "character+separate combining character" does not end up referring to the same variable as "character with combining character"
